src/Type.hs

```diff
@@ -96,35 +98,37 @@ typecheck' (U.Main contracts) = Act store <$> traverse (checkContract store cons
           in (prependIfNotEmpty e y) <> duplicatesBy f ys
 
 
-detectCycle :: Act -> Err ()
-detectCycle (Act _ contracts) =
-  () <$ foldValidation doDFS Set.empty (map fst calls)
+-- | Sort contracts topologically so there are no backward edges in
+-- the constructor call graph. Throw an error if a cycle is detected.
+topologicalSort :: Act -> Err Act
+topologicalSort (Act store contracts) =
+  Act store . reverse . map snd . OM.assocs <$> foldValidation doDFS OM.empty (map fst calls)
```

Comment:
maybe just a little note somewhere that a reverse post ordering of the vertices visited by the DFS is the same as a topological sort? this was a little non obvious to me at first

Reply:
super clean implementation though!

Reply:
done!